### PR TITLE
Added lists/arrays/hash list concepts for Typescript #295

### LINKED ISSUE #295: Add TypeScript Lists/Arrays/Hashes
# Description

Lists/Arrays/Hashes for TypeScript aren't presently on Code Thesaurus. This issue is to add that concept to the system.

# Requirements

* Copy `web/thesauruses/_meta/lists.json` to the right TypeScript directory in `web/thesauruses/`
* Edit the lists.json file to accurately reflect the latest major version of TypeScript
* Delete the `"name":` fields before your PR
* Add in `"comment":` if you wish to explain something in the code
* Add in `"not-implemented": "true"` if the concept doesn't exist in the language. You can then leave `"code":` blank.

# Claiming This Issue

Please leave a comment below so I can assign you to it. This is to prevent multiple people from working on the same issue. 

Check out [CONTRIBUTING.md](https://github.com/codethesaurus/codethesaur.us/blob/main/CONTRIBUTING.md) for more information on how to contribute. 

Check out [https://docs.codethesaur.us/thesaurus/](https://docs.codethesaur.us/thesaurus/) for information on how Code Thesaurus works and how to add to the thesauruses.

# Need Help?

Leave a comment on here with your question, or feel free to ask [@geekygirlsarah](https://twitter.com/geekygirlsarah) on Twitter. You can also email coreteam@codethesaur.us too.

# Thanks!

We really value and appreciate the additions people make to Code Thesaurus, no matter how large or small. So thank you for taking the time to work on this!

### PR DESCRIPTION
**Pull request by abhishek7997**
## What GitHub issue does this PR apply to?

Resolves #295 

## What changed and why?

lists.json file is added under the typescript directory, with the required changes.

## Checklist

- [X] I claimed any associated issue(s) and they are not someone else's
- [X] I have looked at documentation to ensure I made any revisions correctly
- [X] I tested my changes locally to ensure they work
- [ ] (If editing Django) I have added or edited any appropriate unit tests for my changes

## Any additional comments or things to be aware of while reviewing?

Please clarify the difference between "Do two lists match every element?" and "Do two lists contain all the same items?".
I think they mean the same thing.

### CODE REVIEW

File: web/thesauruses/typescript/lists.json [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+      "not-implemented": true
+    },
+    "delete_ordered_mutable_list": {
+      "not-implemented": true
+    },
+    "name_of_unordered_mutable_list": {
+      "code": "Array"
+    },
+    "create_a_unordered_mutable_list": {
+      "code": "Syntax: let arrayName: type[];\nExamples: \nlet scores : (string | number)[];\nlet nums : number[] = [12.34,1,2,3];\nlet arr : Array<any> = [1,2,3,'hi'];\nlet arr : Array<number|boolean> = [1, true, 2];"
```
Comment (geekygirlsarah) [inline web/thesauruses/typescript/lists.json]: The `code` lines should be for actual compileable code. I get what you're doing, but if you want to explain things or mention examples, they should go in a `comment` and leave the `code` for the part that's valid TypeScript. The exception  might be if there's more than one way to do it, then maybe differentiate them in either `comment` or with a comment within the `code`.

**Review by geekygirlsarah — CHANGES_REQUESTED** [review]

File: web/thesauruses/typescript/lists.json [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+    },
+    "erase_element_in_middle_of_unordered_mutable_list": {
+      "code": "arrayName.splice(index,1)"
+    },
+    "swap_elements_in_unordered_mutable_list": {
+      "code": "[array[index1], array[index2]] = [array[index2], array[index1]];"
+    },
+    "delete_unordered_mutable_list": {
+      "code": "arrayName.length = 0",
+      "comment": "In strict mode, use of delete keyword on simple data is prohibited"
```
Comment (geekygirlsarah) [inline web/thesauruses/typescript/lists.json]: You refer to the delete keyword can't be used in strict mode, but what about the un-strict mode? Is there a delete keyword for that? If there's two ways to do it, you can show both, just comment what's going on.

File: web/thesauruses/typescript/lists.json [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+    },
+    "name_of_ordered_immutable_list": {
+      "code": "",
```
Comment (geekygirlsarah) [inline web/thesauruses/typescript/lists.json]: For `not-implemented` lines, could you delete the empty `code` so it cleans it up a bit?

File: web/thesauruses/typescript/lists.json (hunk 2 above)
Comment (abhishek7997) [reply web/thesauruses/typescript/lists.json]: Ok

File: web/thesauruses/typescript/lists.json [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+    },
+    "convert_list_to_string": {
+      "code": "array.toString()"
+    },
+    "concatenate_two_lists": {
+      "code": "let newArray = array1.concat(array2,array3,...)\nreturn newArray"
+    },
+    "split_list_at_index": {
+      "code": "let arr = [ ... ];\nlet part1 = arr.slice(0,index);\nlet part2 = arr.slice(index);",
+      "comment": "There is no inbuilt function that takes an index and returns two arrays"
```
Comment (geekygirlsarah) [inline web/thesauruses/typescript/lists.json:328]: Typically, if there's not a built-in way to do something, then I'd just leave it as a `not-implemented: true` type thing. However, this now has me wondering if it's a simple solution (like your two lines here), maybe I should allow these? I wanted Code Thesaurus to really have simple, built-in replacements but maybe this is fine. 🤔

**Review by geekygirlsarah — CHANGES_REQUESTED** [review]

**Comment by geekygirlsarah** [comment]
This is pretty good! I appreciate the work. Make the minor change and I can probably merge it in. (And thanks for giving me something to think about with whether to allow only built-in functions or small, sort solutions to similar things.)

**Comment by abhishek7997** [comment]
> This is pretty good! I appreciate the work. Make the minor change and I can probably merge it in. (And thanks for giving me something to think about with whether to allow only built-in functions or small, sort solutions to similar things.)

Newline '\n' is not working in comment.
image

"create_a_unordered_mutable_list": {
      "code": "let arrayName: type[];",
      "comment": "Examples: let scores : (string | number)[];\nlet nums : number[] = [12.34,1,2,3];\nlet arr : Array<any> = [1,2,3,'hi'];\nlet arr : Array<number|boolean> = [1, true, 2];"
    },

File: web/thesauruses/typescript/lists.json (hunk 2 above)
Comment (abhishek7997) [reply web/thesauruses/typescript/lists.json]: I tried in unstrict mode, but I found that the entire array cannot be deleted using the 'delete' keyword, another way is to set arrayName to undefined.

**Comment by abhishek7997** [comment]
Please check my updated PR. @geekygirlsarah

**Comment by abhishek7997** [comment]
Please check my updated PR. @geekygirlsarah

**Comment by geekygirlsarah** [comment]
> 
> 
> Please check my updated PR. @geekygirlsarah

I'm just one maintainer and this is my side project. I also have a lot of other life activities going on. I'm addressing the PRs and comments in the order they come in. I'll get to yours, just please be patient.

**Comment by abhishek7997** [comment]
> > Please check my updated PR. @geekygirlsarah
> 
> I'm just one maintainer and this is my side project. I also have a lot of other life activities going on. I'm addressing the PRs and comments in the order they come in. I'll get to yours, just please be patient.

Okay.

**Comment by geekygirlsarah** [comment]
> Newline '\n' is not working in comment. image

Thanks for pointing this out! I made issue #362 to fix it!

You can leave in the `\n` and it'll eventually work when the bug is fixed. (If someone doesn't grab it, maybe I'll do it this week as it shoudl be simple.)

File: web/thesauruses/typescript/lists.json (hunk 1 above)
Comment (geekygirlsarah) [reply web/thesauruses/typescript/lists.json]: You still have uncompileable code here. Please keep `code` to just `let arrayName: type[];` and use examples or other comments in a `comment` line you need to add.

**Review by geekygirlsarah — CHANGES_REQUESTED** [review]

File: web/thesauruses/typescript/lists.json (hunk 1 above)
Comment (geekygirlsarah) [reply web/thesauruses/typescript/lists.json]: I'm still waiting on this part to get fixed. I think I can merge it in if you wrap this up.

Comment (abhishek7997) [reply web/thesauruses/typescript/lists.json]: @geekygirlsarah I have already made the changes. Please check my latest commit.

image

image

**Comment by abhishek7997** [comment]
@geekygirlsarah I need some help. I think I may have pushed too many commits. I got confused as to what branch I should push from. Can you please delete/merge those last 4 commits?

**Comment by abhishek7997** [comment]
@geekygirlsarah Please check my PR. From my end everything looks alright.

**Comment by geekygirlsarah** [comment]
> 
> 
> @geekygirlsarah I need some help. I think I may have pushed too many commits. I got confused as to what branch I should push from. Can you please delete/merge those last 4 commits?

I can't remove commits as they are on your branch and I don't have access to it. Only you can do it and would probably have to force-push it after resetting or re-committing or something. I'm not worried about it as the PR says ultimately you're changing one file and the file looks correct.

I'll merge it in. Thank you!

**Comment by abhishek7997** [comment]
> > @geekygirlsarah I need some help. I think I may have pushed too many commits. I got confused as to what branch I should push from. Can you please delete/merge those last 4 commits?
> 
> I can't remove commits as they are on your branch and I don't have access to it. Only you can do it and would probably have to force-push it after resetting or re-committing or something. I'm not worried about it as the PR says ultimately you're changing one file and the file looks correct.
> 
> I'll merge it in. Thank you!

Okay. Thank s.

**Pull request merged**